### PR TITLE
update postcss value parser for @import and at rule

### PR DESCRIPTION
**Pull request by ad1992**
fixes https://github.com/prettier/prettier/issues/4313.  WIP on fixing the specs.

### CODE REVIEW

File: src/language-css/printer-postcss.js [hunk 1]
```diff
@@ -582,7 +582,19 @@ function genericPrint(path, options, print) {
           isNextForKeyword ||
           isNextEachKeyword
         ) {
-          parts.push(line);
+          if (atRuleAncestorNode && atRuleAncestorNode.name === "import") {
```
Comment (ad1992) [inline src/language-css/printer-postcss.js]: @evilebottnawi I have added some more specs and added more checks in here but I think these checks are too many and the `css_comments` is still different from the original one. Can you tell me how can I make it better ?

File: src/language-css/printer-postcss.js [hunk 2]
```diff
@@ -593,7 +605,13 @@ function genericPrint(path, options, print) {
       if (isControlDirective) {
         return group(indent(concat(parts)));
       }
-
+      if (
+        atRuleAncestorNode &&
+        atRuleAncestorNode.name === "import" &&
+        node.groups[0].value === "url"
```
Comment (evilebottnawi) [inline src/language-css/printer-postcss.js]: Improct can be without `url`function

File: src/language-css/printer-postcss.js (hunk 1 above)
Comment (evilebottnawi) [reply src/language-css/printer-postcss.js]: Alright with comments, don't touch their right now

File: src/language-css/parser-postcss.js [hunk 3]
```diff
@@ -423,7 +423,11 @@ function parseNestedCSS(node) {
           };
         }
 
-        node.params = parseMediaQuery(params);
+        if (lowercasedName === "import") {
+          node.params = parseValue(params);
+        } else {
+          node.params = parseMediaQuery(params);
+        }
```
Comment (evilebottnawi) [inline src/language-css/parser-postcss.js]: Move this here https://github.com/ad1992/prettier/blob/e84c9949924e7b1046c07dff63a0c914abc0f6d7/src/language-css/parser-postcss.js#L394

File: src/language-css/parser-postcss.js [hunk 4]
```diff
@@ -375,6 +375,18 @@ function parseNestedCSS(node) {
         return node;
       }
 
+      if (lowercasedName === "import") {
+        if (params.includes("#{")) {
```
Comment (evilebottnawi) [inline src/language-css/parser-postcss.js]: Don't use this, `postcss-values-parser` works great with `#{$var}`

File: tests/css_atrule/__snapshots__/jsfmt.spec.js.snap [hunk 5]
```diff
@@ -2288,6 +2295,8 @@ $dir: "fonts";
 @import url("very-very-very-very-very-very-very-very-very-very-long-name.css")
   projection,
   tv;
+@import url("very-very-very-very-very-very-very-very-very-very-long-name.css")
+projection tv;
```
Comment (evilebottnawi) [inline tests/css_atrule/__snapshots__/jsfmt.spec.js.snap]: Looks like problem with indent

File: src/language-css/printer-postcss.js [hunk 6]
```diff
@@ -582,7 +582,19 @@ function genericPrint(path, options, print) {
...
+          if (atRuleAncestorNode && atRuleAncestorNode.name === "import") {
+            if (iNextNode.open) {
+              parts.push(iNextNode.open.raws.before.replace(/\s\s+/g, " "));
+            } else if (iNode.close || (iNode.group && iNode.group.close)) {
+              parts.push(line);
+            } else if (iNextNode && iNextNode.raws) {
+              parts.push(iNextNode.raws.before.replace(/\s\s+/g, " "));
+            } else if (iNode && iNode.raws) {
+              parts.push(iNode.raws.after.replace(/\s\s+/g, " "));
+            }
```
Comment (evilebottnawi) [inline src/language-css/printer-postcss.js]: Looks very crazy, we should:
1. Search other solution
2. If no solution found add comments about each `if`

Even I can not understand what is happening here :confused:

Comment (ad1992) [reply src/language-css/printer-postcss.js]: @evilebottnawi I added this code when I tried fixing comments and even I didn't like this code . But I am removing it now as you said to keep comments as it is and adding url check https://github.com/prettier/prettier/pull/4435/commits/ff3d67ff2d49e24351b8b3aa490edcac8cf45c9e

File: src/language-css/parser-postcss.js (hunk 4 above)
Comment (ad1992) [reply src/language-css/parser-postcss.js]: screen shot 2018-05-12 at 10 30 21 pm

It was adding a comma why I added this check here @evilebottnawi 
should I remove this check from here and try fixing it in `printer-postcss.js file` ?

Comment (evilebottnawi) [reply src/language-css/parser-postcss.js]: @ad1992 it is very strange, here only `@import` check, code around `media` below and we have check on `#{`

Comment (ad1992) [reply src/language-css/parser-postcss.js]: @evilebottnawi my bad. this is fixed. 
Though all the test cases except for comments pass now but there are still some issues on how the postcss value parser behaves when having `,` in import statement .  i am working on fixing it.
For example testcase like
`@import url("very-very-very-very-very-very-very-very-very-very-long-name.css") projection,tv,tv1;
`
output of media query parser
```css
@import url("very-very-very-very-very-very-very-very-very-very-long-name.css")
  projection,
  tv,
  tv1;
```
Output of post css value parser
```css
@import url("very-very-very-very-very-very-very-very-very-very-long-name.css")
  projection,
  tv,  tv1;
```

File: src/language-css/printer-postcss.js [hunk 7]
```diff
@@ -593,7 +593,14 @@ function genericPrint(path, options, print) {
       if (isControlDirective) {
         return group(indent(concat(parts)));
       }
-
+      if (
+        atRuleAncestorNode &&
+        atRuleAncestorNode.name === "import" &&
+        node.groups[0].value === "url" &&
+        node.groups.length === 2
```
Comment (evilebottnawi) [inline src/language-css/printer-postcss.js]: Why two? Can you add comments about this?

Comment (ad1992) [reply src/language-css/printer-postcss.js]: @evilebottnawi Indent is not needed for import url when node has two groups

File: src/language-css/printer-postcss.js [hunk 8]
```diff
@@ -593,6 +593,13 @@ function genericPrint(path, options, print) {
       if (isControlDirective) {
         return group(indent(concat(parts)));
       }
+
+      /**
+      * Indent is not needed for import url when node has two groups
+      * when type is value-comma_group
+      * example @import url("abc") projection,tv
+      */
```
Comment (evilebottnawi) [inline src/language-css/printer-postcss.js]: Good comment, don't use `/*` and `*/` it is for JSDoc comments, use simple `//`

**Pull request merged**